lib/Sema/CSSimplify.cpp

```diff
@@ -1836,20 +1835,18 @@ ConstraintSystem::matchTypesBindTypeVar(
   }
 
   // Disallow bindings of noescape functions to type variables that
-  // represent an opened archetype. If we allowed this it would allow
-  // the noescape function to potentially escape.
-  if (auto *fnTy = type->getAs<FunctionType>()) {
-    if (fnTy->isNoEscape() && typeVar->getImpl().getGenericParameter()) {
-      if (shouldAttemptFixes()) {
-        auto *fix = MarkExplicitlyEscaping::create(
-            *this, getConstraintLocator(locator));
-        if (recordFix(fix))
-          return getTypeMatchFailure(locator);
-
-        // Allow no-escape function to be bound with recorded fix.
-      } else {
+  // represent an opened generic parameter. If we allowed this it would
+  // allow the noescape function to potentially escape.
+  if (type->isNoEscape() && typeVar->getImpl().getGenericParameter()) {
```

Comment:
Nit: Comments in this section are out of date now since it's not only functions but tuple types too.

Reply:
Good point, I'll fix that.

Reply:
BTW, I think these checks would be cleaner if we added a flag, like `TVO_CanBindNoEscape` for type variables that can have noescape types, and handled it like `TVO_CanBindInOut`.

Reply:
I agree, that would make more sense.